src/main/java/org/springframework/integration/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -0,0 +1,481 @@
+import com.gs.collections.impl.block.function.checked.CheckedFunction;
+import com.gs.collections.impl.factory.Lists;
+import com.gs.collections.impl.factory.Multimaps;
+import com.gs.collections.impl.list.mutable.FastList;
+import kafka.common.ErrorMapping;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.context.SmartLifecycle;
+import org.springframework.integration.kafka.core.BrokerAddress;
+import org.springframework.integration.kafka.core.ConnectionFactory;
+import org.springframework.integration.kafka.core.ConsumerException;
+import org.springframework.integration.kafka.core.FetchRequest;
+import org.springframework.integration.kafka.core.KafkaMessage;
+import org.springframework.integration.kafka.core.KafkaMessageBatch;
+import org.springframework.integration.kafka.core.KafkaTemplate;
+import org.springframework.integration.kafka.core.Partition;
+import org.springframework.integration.kafka.core.Result;
+import org.springframework.util.Assert;
+
+/**
+ * @author Marius Bogoevici
+ */
+public class KafkaMessageListenerContainer implements SmartLifecycle {
+
+	private static final Log log = LogFactory.getLog(KafkaMessageListenerContainer.class);
+
+	public static final Function<Map.Entry<Partition, ?>, Partition> keyFunction = Functions.getKeyFunction();
+
+	private final GetOffsetForPartitionFunction getOffset = new GetOffsetForPartitionFunction();
+
+	private final PartitionToLeaderFunction getLeader = new PartitionToLeaderFunction();
+
+	private final Function<Partition, Partition> passThru = Functions.getPassThru();
+
+	private final LaunchFetchTaskProcedure launchFetchTask = new LaunchFetchTaskProcedure();
+
+	private final Object lifecycleMonitor = new Object();
+
+	private final KafkaTemplate kafkaTemplate;
+
+	private final Partition[] partitions;
+
+	public boolean autoStartup = true;
+
+	private Executor fetchTaskExecutor;
+
+	private Executor adminTaskExecutor = Executors.newSingleThreadExecutor();
+
+	private int concurrency = 1;
+
+	private volatile boolean running = false;
+
+	private long interval = 100L;
+
+	private int maxFetchSizeInBytes = 10000;
+
+	private MessageListener messageListener;
+
+	private volatile OffsetManager offsetManager;
+
+	private ConcurrentMap<Partition, Long> fetchOffsets;
+
+	private ConcurrentMessageListenerDispatcher messageDispatcher;
+
+	private final MutableMultimap<BrokerAddress, Partition> partitionsByBrokerMap = Multimaps.mutable.set.with();
+
+	public KafkaMessageListenerContainer(ConnectionFactory connectionFactory, Partition[] partitions) {
+		Assert.notNull(connectionFactory, "A connection factory must be supplied");
+		Assert.notEmpty(partitions, "A list of partitions must be provided");
+		this.kafkaTemplate = new KafkaTemplate(connectionFactory);
+		this.partitions = partitions;
+	}
+
+	public KafkaMessageListenerContainer(final ConnectionFactory connectionFactory, String[] topics) {
```

Comment:
Fixed. I had to move the second set of assertions inside `getPartitionsForTopics` to comply with the requirement that `this()` must be called first.